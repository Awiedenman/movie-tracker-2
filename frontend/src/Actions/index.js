export const fetchInitialMovies = movies => ({
  type: 'FETCH_INITIAL_MOVIES',
  movies
});

export const userLogin = user => ({
  type: 'USER_LOGIN',
  id: user.id,
  name: user.name,
  loggedIn: user.loggedIn
});

export const userSignUp = newUserData => ({
  type: 'USER_SIGN_UP',
  id: newUserData.id
});

export const userSignOut = () => ({
  type: 'USER_SIGN_OUT'
});

export const clearUserFavorites = () => ({
  type: 'CLEAR_USER_FAVORITES'
});

export const addFavorite = movie => ({
  type: 'ADD_FAVORITE',
  favorite: movie
});

export const removeFavorite = () => ({
  type: 'REMOVE_FAVORITE'
});

export const addExistingFavorites = existingUserFavorites => {
  return {
    type: 'ADD_EXISTING_FAVORITES',
    favorites : existingUserFavorites
  };
};