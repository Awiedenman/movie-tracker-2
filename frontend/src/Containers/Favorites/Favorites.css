.error-message {
  color: var(--primary-white);
  font-size: 2.5rem;
  font-weight: 400;
  left: 50%;
  position: absolute;
  text-align: center;
  top: 50%;
  transform: translate(-50%, -50%);
}