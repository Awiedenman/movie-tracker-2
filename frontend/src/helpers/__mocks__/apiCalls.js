export const postUserFavorites = jest.fn();
export const movieFetch = jest.fn();
export const fetchUserFavorites = jest.fn().mockImplementation(() =>
  Promise.resolve({
    ok: true,
    data: []
  }));
export const deleteUserFavorite = jest.fn().mockImplementation(() =>
  Promise.resolve({
    ok: true,
    json: () => Promise.resolve({status: 'success', message: 'One row was deleted'})
  }));