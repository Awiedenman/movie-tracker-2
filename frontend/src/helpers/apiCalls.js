import { API_KEY } from '../api-key';

export const movieFetch = async () => {
  const url = `https://api.themoviedb.org/3/list/1?api_key=${API_KEY}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw Error('failed fetch request');
  }

  return await response.json();
};

export const userLoginRequest = async (email, password) => {
  const response = await fetch('http://localhost:3000/api/users',  {
    method: 'POST',
    body: JSON.stringify({ email,  password }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (!response.ok) {
    throw Error('Email and Password do not match');
  }

  return await response.json();
};

export const userSignUpRequest = async (name, email, password) =>  {
  const response = await fetch('http://localhost:3000/api/users/new', {
    method: 'POST',
    body: JSON.stringify({ name, email, password }),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (!response.ok) {
    throw Error('Sorry, email has already been used');
  }

  return await response.json();
};

export const postUserFavorites = async (movie, userId) => {
  const { id, title, image, releaseDate, overview, average } = movie;

  const favoriteToPost = {
    movie_id: id,
    user_id: userId,
    title: title,
    poster_path: image,
    release_date: releaseDate,
    vote_average: average,
    overview: overview
  };

  const response = await fetch('http://localhost:3000/api/users/favorites/new', {
    method: 'POST',
    body: JSON.stringify(favoriteToPost),
    headers: {
      'Content-Type': 'application/json'
    }
  });

  if (!response.ok) {
    throw Error('Sorry, there was problem saving you favorite. Please try again later.');
  }

  return response.json();
};

export const fetchUserFavorites = async userId => {
  const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites`);

  if (!response.ok) {
    throw Error('Sorry, we could not retrieve you favorites at this time');
  }

  return await response.json();
};

export const deleteUserfavorite = async (movieId, userId) => {
  const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites/${movieId}`, {
    method: 'POST',
    body: JSON.stringify(userId, movieId),
    headers: {
      'Content-Type': 'application/json'
    }
  });
  if (!response.ok) {
    throw Error('Sorry, we could not remove you favorite at this time.')
  }
  return await response.json();

};