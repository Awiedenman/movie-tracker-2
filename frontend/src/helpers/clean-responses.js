export const cleanInitialMovieRequest = response => {
  const { items } = response;
  return {
    movieList: items
  };
};