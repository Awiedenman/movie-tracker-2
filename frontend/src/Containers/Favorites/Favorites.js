import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Card from '../../Components/Card/Card';

import './Favorites.css';

export class Favorites extends Component {
  render() {
    const { movies, user } = this.props;
    const userFavorites = movies.map(movie =>
      <Card
        movie={movie}
        key={movie.id}
        userId={user.id}
        favorites={movies}
      />
    );

    if (!user.id) {
      return (
        <div className="home-container">
          <h2>Please Sign In/Sign Up to add Favorites</h2>
        </div>
      );
    }

    if (!movies.length) {
      return (
        <div className="home-container">
          <h2>Please Add Favorites</h2>
        </div>
      );
    }

    return (
      <div className="favorite-container">
        {userFavorites}
      </div>
    );
  }
}

Favorites.propTypes = {
  user: PropTypes.object,
  movies: PropTypes.arrayOf(PropTypes.object)
};

export const mapStateToProps = state => ({
  user: state.userInfo,
  movies: state.favorites
});

export default connect(mapStateToProps)(Favorites);
