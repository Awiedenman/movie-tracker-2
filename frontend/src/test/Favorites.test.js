import React from 'react';
import { shallow } from 'enzyme';
import { Favorites, mapStateToProps, mapDispatchToProps } from '../Containers/Favorites/Favorites';
import { fetchUserFavorites } from '../helpers/apiCalls';
import { removeFavorite } from '../Actions';

jest.mock('../helpers/apiCalls.js');

describe('Favorites', () => {
  let wrapper;
  const mockUser = {
    id: 1
  };
  const mockFavorites = [{name: 'thor', id: 2}];
  const mockRemoveFunc = jest.fn();

  const mockMovie = [{name:'thor', id: 1}];
  beforeEach(() => wrapper = shallow(
    <Favorites
      user={mockUser}
      movies={mockMovie}
      favorites={mockFavorites}
      removeFavorite={mockRemoveFunc}
    />
  ));

  describe('mapStateToProps', () => {
    test('should return a props object with a movies array', () => {
      const expected = {
        favorites: mockMovie,
        user: {
          id: 1
        }
      };

      const mockState = {
        favorites: mockMovie,
        userInfo: {
          id: 1
        }
      };

      const actual = mapStateToProps(mockState);

      expect(actual).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mockMovie = { title: 'batman', id: 2 };
    const mockUserId = 1;
    const actionToDispatch = removeFavorite(mockUserId, mockMovie);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.removeFavorite(mockUserId, mockMovie);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });


  test('should match snapshot when no movies present in favorites', () => {
    wrapper = shallow(
      <Favorites
        user={mockUser}
        movies={mockMovie}
        favorites={[]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });


  test('should should call removeFavorite if favorite already in fav array', async () => {
    const mockMovie = { title: 'Nerds', id: 3 };
    const mockUserId = 1;

    await wrapper.instance().toggleFavorite(mockMovie, mockUserId);
    await removeFavorite();
    expect(mockRemoveFunc).toHaveBeenCalled();
  });

  test('should match snapshot when no user present in', () => {
    wrapper = shallow(
      <Favorites
        user={{}}
        movies={mockMovie}
        favorites={[]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
