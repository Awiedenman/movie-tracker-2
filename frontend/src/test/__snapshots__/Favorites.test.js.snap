// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorites={
        Array [
            Object {
              "id": 2,
              "name": "thor",
            },
          ]
    }
    movies={
        Array [
            Object {
              "id": 1,
              "name": "thor",
            },
          ]
    }
    removeFavorite={[Function]}
    user={
        Object {
            "id": 1,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Card
          favorites={
                    Array [
                              Object {
                                "id": 2,
                                "name": "thor",
                              },
                            ]
          }
          movie={
                    Object {
                              "id": 2,
                              "name": "thor",
                            }
          }
          toggleFavorite={[Function]}
          userId={1}
/>,
      ],
      "className": "favorite-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "favorites": Array [
            Object {
              "id": 2,
              "name": "thor",
            },
          ],
          "movie": Object {
            "id": 2,
            "name": "thor",
          },
          "toggleFavorite": [Function],
          "userId": 1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Card
            favorites={
                        Array [
                                    Object {
                                      "id": 2,
                                      "name": "thor",
                                    },
                                  ]
            }
            movie={
                        Object {
                                    "id": 2,
                                    "name": "thor",
                                  }
            }
            toggleFavorite={[Function]}
            userId={1}
/>,
        ],
        "className": "favorite-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "favorites": Array [
              Object {
                "id": 2,
                "name": "thor",
              },
            ],
            "movie": Object {
              "id": 2,
              "name": "thor",
            },
            "toggleFavorite": [Function],
            "userId": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`Favorites should match snapshot when no movies present in favorites 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorites={Array []}
    movies={
        Array [
            Object {
              "id": 1,
              "name": "thor",
            },
          ]
    }
    user={
        Object {
            "id": 1,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h2>
        Please Add Favorites
</h2>,
      "className": "home-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Please Add Favorites",
      },
      "ref": null,
      "rendered": "Please Add Favorites",
      "type": "h2",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h2>
          Please Add Favorites
</h2>,
        "className": "home-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Please Add Favorites",
        },
        "ref": null,
        "rendered": "Please Add Favorites",
        "type": "h2",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`Favorites should match snapshot when no user present in 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorites={Array []}
    movies={
        Array [
            Object {
              "id": 1,
              "name": "thor",
            },
          ]
    }
    user={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h2>
        Please Sign In/Sign Up to add Favorites
</h2>,
      "className": "home-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Please Sign In/Sign Up to add Favorites",
      },
      "ref": null,
      "rendered": "Please Sign In/Sign Up to add Favorites",
      "type": "h2",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h2>
          Please Sign In/Sign Up to add Favorites
</h2>,
        "className": "home-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Please Sign In/Sign Up to add Favorites",
        },
        "ref": null,
        "rendered": "Please Sign In/Sign Up to add Favorites",
        "type": "h2",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
